{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Developing function demo\n",
    "----------------------\n",
    "\n",
    "1. Need a mock gene list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note about below line. Should be unnecessary now. When developing this notebook, `intermine` module was not yet in the `environment.yml` for the Binder-served repo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting intermine\n",
      "  Downloading intermine-1.09.09.tar.gz (63kB)\n",
      "\u001b[K    100% |████████████████████████████████| 71kB 2.2MB/s ta 0:00:01\n",
      "\u001b[?25hBuilding wheels for collected packages: intermine\n",
      "  Running setup.py bdist_wheel for intermine ... \u001b[?25ldone\n",
      "\u001b[?25h  Stored in directory: /home/jovyan/.cache/pip/wheels/b3/bb/a3/9a5aca5755bc817f165a364ce1ff4f9687e9fea14cfcb895a0\n",
      "Successfully built intermine\n",
      "Installing collected packages: intermine\n",
      "Successfully installed intermine-1.9.9\n"
     ]
    }
   ],
   "source": [
    "!pip install intermine"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To do the process represented in the next cell,.i.e., load a particular version of a script from Github and save it to the Binder served repo. (A big reason for doing it this way, without leaving it in the repo, is that it would make multiple copies and best for maintening if everything kept served from one source, I think.)\n",
    "\n",
    "**1) Pasted in the below cell the code below and then executed that cell**\n",
    "\n",
    "    %load https://raw.githubusercontent.com/fomightez/yeastmine/01e84c2704de405ad67b75d352b7fd32c2cf4ae1/make_simulated_yeast_gene_set.py\n",
    "\n",
    "**2) That resulted in loading the particular verison of the script and commented out the load command that had been used.**\n",
    "\n",
    "**3) To save it so that `%run` will work. In the cell where the script just loaded, I added the text below into the top and ran that cell again.**\n",
    "\n",
    "    %%writefile make_simulated_yeast_gene_set.py\n",
    "\n",
    "That gives output of `Writing make_simulated_yeast_gene_set.py` indicating it generated the file on the server where it is being run via binder.  That last step was based on [here](https://stackoverflow.com/a/21034906/8508004).\n",
    "\n",
    "**4) Now**, IN A NEW CELL, **I can run the script with the command below where I can add arguments, in this case specifying size of list to make. This is what I would put in that cell.**\n",
    "\n",
    "    %run make_simulated_yeast_gene_set.py 6918\n",
    "\n",
    "(It seems like there ought to be a better way to do this, i.e., put a script that I need to run with arguments into the notebook and run it,  but I haven't determined it yet. `%load` alone works for functions. Besides being clunky, I think this is an example where the notebook system doesn't reflect how things were done without extra information.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing make_simulated_yeast_gene_set.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile make_simulated_yeast_gene_set.py\n",
    "# %load https://raw.githubusercontent.com/fomightez/yeastmine/01e84c2704de405ad67b75d352b7fd32c2cf4ae1/make_simulated_yeast_gene_set.py\n",
    "#!/usr/bin/env python\n",
    "# make_simulated_yeast_gene_set.py by Wayne Decatur\n",
    "# ver 0.1\n",
    "#\n",
    "#*******************************************************************************\n",
    "# Compatible with both Python 2.7 and Python 3.6 (verified); written initially \n",
    "# in Python 2.7 to hopefully be convertible to Python 3.\n",
    "#\n",
    "# PURPOSE: Makes a simulated yeast gene set of user-determined size and saves\n",
    "# a file of the gene list using the SGD systematic IDs each on individual lines.\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "# Dependencies beyond the mostly standard libraries/modules:\n",
    "# intermine\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "# VERSION HISTORY:\n",
    "# v.0.1. basic working version\n",
    "#\n",
    "\n",
    "#\n",
    "#\n",
    "#\n",
    "# TO RUN:\n",
    "# Example,\n",
    "# Enter on the command line of your terminal, the line\n",
    "#-----------------------------------\n",
    "# make_simulated_yeast_gene_set.py 135\n",
    "#-----------------------------------\n",
    "# Replace `135` with the number of genes you want in your list.\n",
    "#\n",
    "#*******************************************************************************\n",
    "#\n",
    "\n",
    "\n",
    "#*******************************************************************************\n",
    "##################################\n",
    "#  USER ADJUSTABLE VALUES        #\n",
    "\n",
    "##################################\n",
    "#\n",
    "#\n",
    "output_file_name = \"simulated_yeast_gene_list.txt\"\n",
    "\n",
    "stored_gene_data_file_prefix = \"yeastmine_gene_ids_for_creating_random_set\" #\n",
    "# prefix for file name for data to save for subsequent use or to use from\n",
    "# previous runs.\n",
    "\n",
    "progress_bar_use = False # `False` to disable by default because relies on\n",
    "# an uncommon module. Fancier progressbar if set to `True`. If you get an error\n",
    "# concerning `import tqdm`, set this to `False`. Or the necessary\n",
    "# module can be installed on most machines with `pip install tqdm`. On\n",
    "# PythonAnywhere.com, it is `pip install --user tqdm` .\n",
    "\n",
    "#\n",
    "#*******************************************************************************\n",
    "#**********************END USER ADJUSTABLE VARIABLES****************************\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#*******************************************************************************\n",
    "#*******************************************************************************\n",
    "###DO NOT EDIT BELOW HERE - ENTER VALUES ABOVE###\n",
    "\n",
    "import sys\n",
    "import os\n",
    "import argparse\n",
    "import random\n",
    "\n",
    "\n",
    "###---------------------------HELPER FUNCTIONS---------------------------------###\n",
    "\n",
    "\n",
    "def generate_stored_data_file_name(file_name_prefix):\n",
    "    '''\n",
    "    Takes a file name as an argument and returns string for the name of the\n",
    "    output file. The generated name is based on the original file\n",
    "    name.\n",
    "\n",
    "    Specific example\n",
    "    ================\n",
    "    Calling function in June 2016 with\n",
    "        (\"yeastmine_gene_data_for_creating_random_set\")\n",
    "    returns\n",
    "        \"yeastmine_gene_data_for_creating_random_setJune2016.py\"\n",
    "\n",
    "    It also returns month and year to use in the data file text.\n",
    "    '''\n",
    "    # prepare to timestamp the name of file to save\n",
    "    import datetime\n",
    "    now = datetime.datetime.now()\n",
    "    # see http://www.saltycrane.com/blog/2008/06/how-to-get-current-date-and-time-in/\n",
    "    # for formatting options simplified\n",
    "    return file_name_prefix + now.strftime(\"%B%Y\") +\".py\", now.strftime(\n",
    "        \"%B\"), now.strftime(\"%Y\")\n",
    "\n",
    "def generate_output_file(provided_text):\n",
    "    '''\n",
    "    function text and saves it as a text file\n",
    "    '''\n",
    "    name_of_file_to_save = output_file_name\n",
    "    data_file_stream = open(name_of_file_to_save , \"w\")\n",
    "    data_file_stream.write(provided_text.rstrip('\\r\\n')) #rstrip to remove trailing newline\n",
    "    # from http://stackoverflow.com/questions/275018/how-can-i-remove-chomp-a-newline-in-python\n",
    "    data_file_stream.close()\n",
    "    sys.stderr.write(\n",
    "        \"\\nGene set composed of {1} genes selected randomly from the {2} genes in the genome saved as '{0}'.\\n\".format(\n",
    "        name_of_file_to_save, number_of_genes_for_list, str(\n",
    "        genes_in_genome)))\n",
    "\n",
    "def list2text(a_list):\n",
    "    '''\n",
    "    a function that takes a lists and makes a string where each item in list\n",
    "    is on a new line\n",
    "    '''\n",
    "    return \"\\n\".join(a_list)\n",
    "\n",
    "\n",
    "def check_one_or_greater(value):\n",
    "    ivalue = int(value)\n",
    "    if ivalue < 1:\n",
    "         raise argparse.ArgumentTypeError(\n",
    "            \"{0} is an invalid value for the number of genes to put in the simulated set; provide an integer that is one or greater.\".format(value))\n",
    "    return ivalue\n",
    "\n",
    "\n",
    "def header_text_note_maker(month, year):\n",
    "    '''\n",
    "    function generates text that puts description(header) for list entries in\n",
    "    gene_data_list\n",
    "\n",
    "    month and year are provided from the calling function to make header better\n",
    "    '''\n",
    "    header_text = \"# These were the identifiers for all the \" + str(\n",
    "        genes_in_genome) +\" yeast genes in YeastMine \"+ month + \" \" + year +\".\\n# The  YeastMine `secondaryIdentifier` (equivalent to SGD's `Systematic Name` identifier) for each gene are as follows:\\n\"\n",
    "    return header_text\n",
    "\n",
    "def generate_data_file(the_data_list):\n",
    "    '''\n",
    "    function takes gene data list from YeastMine and saves it as script-like file\n",
    "    '''\n",
    "\n",
    "    stored_gene_data_file, month, year  = generate_stored_data_file_name(\n",
    "        stored_gene_data_file_prefix) # see next line for info on month & year.\n",
    "    # Month and year are first used in the function just called and so instead\n",
    "    # of calling that import and commands again, just passing back.\n",
    "\n",
    "\n",
    "    data_file_stream = open(stored_gene_data_file , \"w\")\n",
    "    data_file_stream.write(header_text_note_maker(month, year))\n",
    "    data_file_stream.write('gene_data_list = ' + repr(gene_data_list ) + \"\\n\")\n",
    "    data_file_stream.close()\n",
    "    sys.stderr.write( \"\\b \\nGene ids for the \" + str(\n",
    "    genes_in_genome) +\" yeast genes stored for posible future use.\\n\") # `\\b ` in the\n",
    "    # deletes the spinner, if was there; it is important that some text be\n",
    "    # after the backspace because backspace and new line alone doesn't write\n",
    "    # over the spinner otherwise.\n",
    "\n",
    "\n",
    "\n",
    "###--------------------------END OF HELPER FUNCTIONS---------------------------###\n",
    "###--------------------------END OF HELPER FUNCTIONS---------------------------###\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#*******************************************************************************\n",
    "###-----------------for parsing command line arguments-----------------------###\n",
    "parser = argparse.ArgumentParser(prog='make_simulated_yeast_gene_set.py ',description=\"make_simulated_yeast_gene_set.py  \\\n",
    "    Makes a simulated yeast gene set of user-determined size and saves a file \\\n",
    "    of the gene list using the SGD's systematic IDs, each on individual lines.\\\n",
    "    **** Script by Wayne Decatur (fomightez @ github) ***\")\n",
    "parser.add_argument(\"Number\", help=\"Number of genes to have in the produced \\\n",
    "    gene set.\", type=check_one_or_greater)\n",
    "#I would also like trigger help to display if no arguments provided because need at least one input file\n",
    "if len(sys.argv)==1:    #from http://stackoverflow.com/questions/4042452/display-help-message-with-python-argparse-when-script-is-called-without-any-argu\n",
    "    parser.print_help()\n",
    "    sys.exit(1)\n",
    "args = parser.parse_args()\n",
    "number_of_genes_for_list = args.Number\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "###-----------------Actual Main portion of script---------------------------###\n",
    "\n",
    "# Use previously saved list of yeast genes, or if none exists contact YeastMine\n",
    "# and get a list of the genes\n",
    "\n",
    "# To check for a file that begins with the prefix make a list of the files in\n",
    "# the folder with the script and check if one of them begins with the prefix text.\n",
    "# Then use that to get name if any match by taking first one because there never\n",
    "# will be more than one unless debugging. Set to something that couldn't\n",
    "# possibly match anything in folder otherwise so can use\n",
    "# `os.path.isfile(stored_gene_data_file)` without having to first check if `stored_gene_data_file` id defined.\n",
    "# Line below based on Marc L's answer at http://stackoverflow.com/questions/15312953/choose-a-file-starting-with-a-given-string\n",
    "stored_gene_data_file_possible_list = [filename for filename in os.listdir('.') if (filename.startswith(stored_gene_data_file_prefix)) and (not filename.endswith('.pyc'))] # the `not filename.endswith('pyc')` is to exclude getting name of compiled version\n",
    "if stored_gene_data_file_possible_list:\n",
    "    stored_gene_data_file = stored_gene_data_file_possible_list[0]\n",
    "else:\n",
    "    stored_gene_data_file = \"DOESN'T EXIST and this is an impossible name to never match a file\"\n",
    "\n",
    "if os.path.isfile(stored_gene_data_file) :\n",
    "    # If data for conversion table already saved, used that to make dictionary.\n",
    "    # The reason the dictionary wasn't saved is because it drops a lot of the\n",
    "    # gene data information and the additional details saved in the gene data\n",
    "    # list may later help in  sorting if gene is the same or different, etc..\n",
    "    #\n",
    "    #import the saved data list\n",
    "    import importlib\n",
    "    imported_data = importlib.import_module(stored_gene_data_file[:-3], package=None) # `[:-3]` part is because don't want `.py` part of file name when calling module\n",
    "    # instead of using `imported_data.gene_data_list` assign variable shorter name.\n",
    "    gene_data_list_text = imported_data.gene_data_list\n",
    "\n",
    "    # Now almost ready to make the list\n",
    "    # set up list\n",
    "    syst_id_list = []\n",
    "\n",
    "    # Make the list\n",
    "    sys.stderr.write(\"Using stored genes data previously acquired from YeastMine, stored as `\" + stored_gene_data_file +\"`, to make a list of yeast genes.\\n(Rename or delete that file if you prefer the script try to retrieve the pertinent information from YeastMine at this time.)...\" )\n",
    "    genes_processed = 0\n",
    "    for each_gene_id in gene_data_list_text:\n",
    "        syst_id_list.append(each_gene_id)\n",
    "        genes_processed += 1\n",
    "        # no point having progress bar because so fast with data already saved in a file so no call to YeastMine needed.\n",
    "        if (genes_processed  % 1000 == 0):\n",
    "                sys.stderr.write(\".\")\n",
    "\n",
    "else:\n",
    "\n",
    "    # To run your query\n",
    "    # to use it you will require the intermine python client.\n",
    "    # To install the client, run the following command from a terminal:\n",
    "    #\n",
    "    #     sudo easy_install intermine\n",
    "    #\n",
    "    # For further documentation you can visit:\n",
    "    #     http://intermine.readthedocs.org/en/latest/web-services/\n",
    "\n",
    "    # The following two lines will be needed in every python script:\n",
    "    from intermine.webservice import Service\n",
    "    service = Service(\"https://yeastmine.yeastgenome.org:443/yeastmine/service\") #seems current as of January 2018 from the YeastMine site example, I had also prior to this change in the script, run on my machine ` sudo easy_install intermine --upgrade`\n",
    "    query = service.new_query(\"SequenceFeature\")\n",
    "    query.add_view(\n",
    "        \"primaryIdentifier\", \"featureType\", \"secondaryIdentifier\", \"description\",\n",
    "        \"sgdAlias\", \"symbol\"\n",
    "    )\n",
    "    query.add_constraint(\"featureType\", \"=\", \"telomerase_RNA_gene\", code = \"Z\")\n",
    "    query.add_constraint(\"qualifier\", \"IS NULL\", code = \"W\")\n",
    "    query.add_constraint(\"qualifier\", \"!=\", \"Dubious\", code = \"V\")\n",
    "    query.add_constraint(\"status\", \"=\", \"Active\", code = \"U\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"transposable_element_gene\", code = \"S\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"telomeric_repeat\", code = \"R\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"telomere\", code = \"Q\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"tRNA_gene\", code = \"P\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"snoRNA_gene\", code = \"O\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"snRNA_gene\", code = \"N\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"LTR_retrotransposon\", code = \"M\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"rRNA_gene\", code = \"L\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"pseudogene\", code = \"K\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"not physically mapped\", code = \"J\")\n",
    "    #query.add_constraint(\"featureType\", \"=\", \"not in systematic sequence of S288C\", code = \"I\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"ncRNA_gene\", code = \"H\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"long_terminal_repeat\", code = \"G\")\n",
    "    #query.add_constraint(\"featureType\", \"=\", \"centromere\", code = \"F\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"Y_prime_element\", code = \"E\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"X_element\", code = \"D\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"X_element_combinatorial_repeat\", code = \"C\")\n",
    "    #query.add_constraint(\"featureType\", \"=\", \"ARS\", code = \"B\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"ORF\", code = \"A\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"blocked_reading_frame\", code = \"X\")\n",
    "    query.add_constraint(\"featureType\", \"=\", \"intein_encoding_region\", code = \"Y\")\n",
    "    query.add_constraint(\"organism.name\", \"=\", \"Saccharomyces cerevisiae\", code = \"T\")\n",
    "    #query.set_logic(\"(A or B or C or D or E or F or G or H or I or J or K or L or M or N or O or P or Q or R or S or X or Y or Z) and T and U and (V or W)\")\n",
    "    query.set_logic(\"(A or C or D or E or G or H or J or K or L or M or N or O or P or Q or R or S or X or Y or Z) and T and U and (V or W)\")\n",
    "\n",
    "\n",
    "    # Now almost ready to make the list.\n",
    "    # set up the list\n",
    "    syst_id_list = []\n",
    "\n",
    "    # prepare a list to store the information for each gene to ultimately save as a file\n",
    "    gene_data_list = []\n",
    "\n",
    "\n",
    "    # Make the list\n",
    "    sys.stderr.write(\"Please wait. Acquiring id information on every gene reported at YeastMine to ultimately make a subset containing the user-requested number...\")\n",
    "    none_instances = 0\n",
    "    if progress_bar_use:\n",
    "        from tqdm import * #see about this module at https://github.com/tqdm/tqdm and https://github.com/noamraph/tqdm\n",
    "        genes_processed = 0\n",
    "        sys.stderr.write(\"\\n\") #go to next line so progress bar doesn't delete earlier text\n",
    "        for row in tqdm(query.rows()):\n",
    "            if row[\"secondaryIdentifier\"] == None:\n",
    "                none_instances += 1\n",
    "                #print row[\"primaryIdentifier\"], row[\"sgdAlias\"],row[\"symbol\"], row[\"description\"], row[\"featureType\"]\n",
    "                # Using that print statement on above line to look into this I found 111 give `None` for row[\"secondaryIdentifier\"]\n",
    "                # and so just used that method here to filter out and not add to lists instead of toying with above since maybe\n",
    "                # some have identifers since when I had filtered more I was down to ~6532 genes ??\n",
    "                '''\n",
    "                S000029636 None None None intein_encoding_region\n",
    "                S000028938 None None None telomeric_repeat\n",
    "                S000028864 None None None telomeric_repeat\n",
    "                S000028941 None None None telomeric_repeat\n",
    "                S000028945 None None None telomeric_repeat\n",
    "                S000028872 None None None telomeric_repeat\n",
    "                S000028949 None None None telomeric_repeat\n",
    "                S000028876 None None None telomeric_repeat\n",
    "                S000028958 None None None telomeric_repeat\n",
    "                S000028883 None None None telomeric_repeat\n",
    "                S000028888 None None None telomeric_repeat\n",
    "                S000028965 None None None telomeric_repeat\n",
    "                S000028966 None None None telomeric_repeat\n",
    "                S000028892 None None None telomeric_repeat\n",
    "                S000028897 None None None telomeric_repeat\n",
    "                S000028971 None None None telomeric_repeat\n",
    "                S000028975 None None None telomeric_repeat\n",
    "                S000028902 None None None telomeric_repeat\n",
    "                S000028979 None None None telomeric_repeat\n",
    "                S000028907 None None None telomeric_repeat\n",
    "                S000028983 None None None telomeric_repeat\n",
    "                S000028984 None None None telomeric_repeat\n",
    "                S000028911 None None None telomeric_repeat\n",
    "                S000028912 None None None telomeric_repeat\n",
    "                S000028913 None None None telomeric_repeat\n",
    "                S000028990 None None None telomeric_repeat\n",
    "                S000028919 None None None telomeric_repeat\n",
    "                S000028920 None None None telomeric_repeat\n",
    "                S000028994 None None None telomeric_repeat\n",
    "                S000028925 None None None telomeric_repeat\n",
    "                S000028998 None None None telomeric_repeat\n",
    "                S000028930 None None None telomeric_repeat\n",
    "                S000028939 None None None X_element\n",
    "                S000028865 None None None X_element\n",
    "                S000028942 None None None X_element\n",
    "                S000028868 None None None X_element\n",
    "                S000028946 None None None X_element\n",
    "                S000028873 None None None X_element\n",
    "                S000028950 None None None X_element\n",
    "                S000028877 None None None X_element\n",
    "                S000028954 None None None X_element\n",
    "                S000028880 None None None X_element\n",
    "                S000028959 None None None X_element\n",
    "                S000028884 None None None X_element\n",
    "                S000028961 None None None X_element\n",
    "                S000028889 None None None X_element\n",
    "                S000028967 None None None X_element\n",
    "                S000028893 None None None X_element\n",
    "                S000028898 None None None X_element\n",
    "                S000028972 None None None X_element\n",
    "                S000028976 None None None X_element\n",
    "                S000028903 None None None X_element\n",
    "                S000028980 None None None X_element\n",
    "                S000028908 None None None X_element\n",
    "                S000028985 None None None X_element\n",
    "                S000028914 None None None X_element\n",
    "                S000028991 None None None X_element\n",
    "                S000028921 None None None X_element\n",
    "                S000028995 None None None X_element\n",
    "                S000028926 None None None X_element\n",
    "                S000028999 None None None X_element\n",
    "                S000028931 None None None X_element\n",
    "                S000028934 None None None X_element\n",
    "                S000029003 None None None X_element\n",
    "                S000028866 None None None X_element_combinatorial_repeat\n",
    "                S000028943 None None None X_element_combinatorial_repeat\n",
    "                S000028869 None None None X_element_combinatorial_repeat\n",
    "                S000028947 None None None X_element_combinatorial_repeat\n",
    "                S000028874 None None None X_element_combinatorial_repeat\n",
    "                S000028951 None None None X_element_combinatorial_repeat\n",
    "                S000028878 None None None X_element_combinatorial_repeat\n",
    "                S000028955 None None None X_element_combinatorial_repeat\n",
    "                S000028885 None None None X_element_combinatorial_repeat\n",
    "                S000028962 None None None X_element_combinatorial_repeat\n",
    "                S000028890 None None None X_element_combinatorial_repeat\n",
    "                S000028968 None None None X_element_combinatorial_repeat\n",
    "                S000028894 None None None X_element_combinatorial_repeat\n",
    "                S000028899 None None None X_element_combinatorial_repeat\n",
    "                S000028973 None None None X_element_combinatorial_repeat\n",
    "                S000028977 None None None X_element_combinatorial_repeat\n",
    "                S000028904 None None None X_element_combinatorial_repeat\n",
    "                S000028981 None None None X_element_combinatorial_repeat\n",
    "                S000028909 None None None X_element_combinatorial_repeat\n",
    "                S000028986 None None None X_element_combinatorial_repeat\n",
    "                S000028915 None None None X_element_combinatorial_repeat\n",
    "                S000028992 None None None X_element_combinatorial_repeat\n",
    "                S000028922 None None None X_element_combinatorial_repeat\n",
    "                S000028996 None None None X_element_combinatorial_repeat\n",
    "                S000028927 None None None X_element_combinatorial_repeat\n",
    "                S000029000 None None None X_element_combinatorial_repeat\n",
    "                S000028932 None None None X_element_combinatorial_repeat\n",
    "                S000028935 None None None X_element_combinatorial_repeat\n",
    "                S000028870 None None None Y_prime_element\n",
    "                S000028952 None None None Y_prime_element\n",
    "                S000028956 None None None Y_prime_element\n",
    "                S000028881 None None None Y_prime_element\n",
    "                S000028886 None None None Y_prime_element\n",
    "                S000028963 None None None Y_prime_element\n",
    "                S000028969 None None None Y_prime_element\n",
    "                S000028895 None None None Y_prime_element\n",
    "                S000028900 None None None Y_prime_element\n",
    "                S000028905 None None None Y_prime_element\n",
    "                S000028988 None None None Y_prime_element\n",
    "                S000028987 None None None Y_prime_element\n",
    "                S000028916 None None None Y_prime_element\n",
    "                S000028917 None None None Y_prime_element\n",
    "                S000028923 None None None Y_prime_element\n",
    "                S000028928 None None None Y_prime_element\n",
    "                S000029001 None None None Y_prime_element\n",
    "                S000028936 None None None Y_prime_element\n",
    "                S000029004 None None None Y_prime_element\n",
    "                '''\n",
    "            else:\n",
    "                genes_processed += 1\n",
    "                syst_id_list.append(row[\"secondaryIdentifier\"])\n",
    "                # Collect the data as a list of lists for possible future use and/or speeding up runs\n",
    "                gene_data_list.append(row[\"secondaryIdentifier\"])\n",
    "\n",
    "    else:\n",
    "        genes_processed = 0\n",
    "        for row in query.rows():\n",
    "            if row[\"secondaryIdentifier\"] == None:\n",
    "                none_instances += 1\n",
    "                #print row[\"primaryIdentifier\"], row[\"sgdAlias\"],row[\"symbol\"], row[\"description\"], row[\"featureType\"]\n",
    "                # SEE ABOVE ABOUT DEBUGGING WITH THAT LINE\n",
    "            else:\n",
    "                genes_processed += 1\n",
    "                syst_id_list.append(row[\"secondaryIdentifier\"])\n",
    "                # Collect the data as a list of lists for possible future use and/or speeding up runs\n",
    "                gene_data_list.append(row[\"secondaryIdentifier\"])\n",
    "            #every 200 lines print\n",
    "            if (genes_processed % 200 == 0):\n",
    "                sys.stderr.write(\".\")\n",
    "\n",
    "    # assign `genes_processed` to something more meaningful\n",
    "    genes_in_genome = genes_processed\n",
    "\n",
    "    #print none_instances # FOR DEBUGGING GENE LIST\n",
    "\n",
    "    # Save the data in gene_data_list for possibe future use or for saving time.\n",
    "    # This will allow storage of the uds. This will serve two\n",
    "    # two purposes:\n",
    "    # 1) This will store the version of YeastMine gene list ids at the time. And\n",
    "    # therefore as gene designations change perhaps in later YeastMine/SGD\n",
    "    # releases, this will have served to capture the data for when the input\n",
    "    # data was originally created. At that time the version of YeastMine/SGD\n",
    "    # should match the generated data.\n",
    "    # 2) This will make the script run faster by eliminating the need to access\n",
    "    #  YeastMine and make on subsequent runs.\n",
    "    # The file will be stored as a \".py\" file for ease but any tecxt editor will\n",
    "    # be able to read it. It may need some manipulation to convery it to a form\n",
    "    # that is easy to read for a human but the data will be there in almost\n",
    "    # readable form.\n",
    "    generate_data_file(gene_data_list)\n",
    "\n",
    "\n",
    "\n",
    "# assign `genes_processed` to something more meaningful\n",
    "genes_in_genome = genes_processed\n",
    "\n",
    "\n",
    "# Now use user-provided number to make a subset from the list of yeast genes\n",
    "subset_of_genes = random.sample(syst_id_list, number_of_genes_for_list)\n",
    "\n",
    "\n",
    "\n",
    "# handle the generating the output list text\n",
    "# Make the list easily made into an output file by separating each with a new line\n",
    "text_to_save = list2text(subset_of_genes)\n",
    "\n",
    "\n",
    "# Save results and signal that to user\n",
    "generate_output_file(text_to_save)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#*******************************************************************************\n",
    "###-***********************END MAIN PORTION OF SCRIPT***********************-###\n",
    "#*******************************************************************************\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using stored genes data previously acquired from YeastMine, stored as `yeastmine_gene_ids_for_creating_random_setFebruary2018.py`, to make a list of yeast genes.\n",
      "(Rename or delete that file if you prefer the script try to retrieve the pertinent information from YeastMine at this time.).........\n",
      "Gene set composed of 6918 genes selected randomly from the 6918 genes in the genome saved as 'simulated_yeast_gene_list.txt'.\n"
     ]
    }
   ],
   "source": [
    "%run make_simulated_yeast_gene_set.py 6918"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SECOND ITEM NEEDED==> The other item needed is a similar gene list\n",
    "\n",
    "Without needing to wait forever, can I shuffle the list but get an instance of the list that is around 63% similar to the one I have when considering rankings. (I actually had more similar real data recently, but by relaxing the similarity, I should be able to get a ranked list that meets those requirements faster.)\n",
    "\n",
    "First I'll need to get the list back into the notebooks's namespace."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "orginal_df = pd.read_table('simulated_yeast_gene_list.txt',header=None, names=['gene'])\n",
    "original_ranked_list = orginal_df['gene'].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then I need the script for assessing similarity, which I can get with `%load https://raw.githubusercontent.com/maslinych/linis-scripts/f9ab60e94417c3bb465f107e93cde80168622e9b/rbo_calc.py`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load https://raw.githubusercontent.com/maslinych/linis-scripts/f9ab60e94417c3bb465f107e93cde80168622e9b/rbo_calc.py\n",
    "def calc_rbo(l1,l2,p):\n",
    "    \"\"\" \n",
    "    Returns RBO indefinite rank similarity metric, as described in:\n",
    "    Webber, W., Moffat, A., & Zobel, J. (2010). \n",
    "    A similarity measure for indefinite rankings. \n",
    "    ACM Transactions on Information Systems.\n",
    "    doi:10.1145/1852102.1852106.\n",
    "    \"\"\"\n",
    "    sl,ll = sorted([(len(l1), l1),(len(l2),l2)])\n",
    "    s, S = sl\n",
    "    l, L = ll\n",
    "    \n",
    "    # Calculate the overlaps at ranks 1 through l \n",
    "    # (the longer of the two lists)\n",
    "    ss = set([])\n",
    "    ls = set([])\n",
    "    overs = {}\n",
    "    for i in range(l):\n",
    "        ls.add(L[i])\n",
    "        if i<s:\n",
    "           ss.add(S[i])\n",
    "        X_d = len(ss.intersection(ls))\n",
    "        d = i+1\n",
    "        overs[d] = float(X_d)\n",
    "    \n",
    "    # (1) \\sum_{d=1}^l (X_d / d) * p^d\n",
    "    sum1 = 0\n",
    "    for i in range(l):\n",
    "        d=i+1\n",
    "        sum1+=overs[d]/d*pow(p,d)\n",
    "    X_s = overs[s]\n",
    "    X_l = overs[l]\n",
    "\n",
    "    # (2) \\sum_{d=s+1}^l [(X_s (d - s)) / (sd)] * p^d\n",
    "    sum2 = 0\n",
    "    for i in range(s,l):\n",
    "        d=i+1\n",
    "        sum2+=(X_s*(d-s)/(s*d))*pow(p,d)\n",
    "\n",
    "    # (3) [(X_l - X_s) / l + X_s / s] * p^l\n",
    "    sum3 = ((X_l-X_s)/l+X_s/s)*pow(p,l)\n",
    "    \n",
    "    # Equation 32. \n",
    "    rbo_ext = (1-p)/p*(sum1+sum2)+sum3\n",
    "    return rbo_ext\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now to run the shuffling a lot of times to see if I can get one that is at least 63% similar."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1h 11min 16s, sys: 2.4 s, total: 1h 11min 19s\n",
      "Wall time: 1h 11min 19s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import random\n",
    "# define values to check\n",
    "similarity_level_achieved = 0 #set to very low to start out\n",
    "min_similarity_neeeded = 0.08 # much lower than I had wanted but completes on the scale of many minutes with that level\n",
    "\n",
    "# BLOCK FOR TESTING\n",
    "#copied_list = list(original_ranked_list) #must copy list first b/c shuffle is done in place\n",
    "#random.shuffle(copied_list)\n",
    "#current_similarity = calc_rbo(original_ranked_list,copied_list,0.98)\n",
    "#print(current_similarity)\n",
    "# END of BLOCK FOR TESTING\n",
    "\n",
    "while (similarity_level_achieved < min_similarity_neeeded):\n",
    "    copied_list = list(original_ranked_list) #must copy list first b/c shuffle is done in place\n",
    "    random.shuffle(copied_list)\n",
    "    #assess the shuffled copy relative original\n",
    "    current_similarity = calc_rbo(original_ranked_list,copied_list,0.98)\n",
    "    if current_similarity > similarity_level_achieved:\n",
    "        similarity_level_achieved = current_similarity\n",
    "# with good list in hand. Save it.\n",
    "df_4save = pd.DataFrame(copied_list, columns=['genes']) # example from https://stackoverflow.com/a/43175477/8508004\n",
    "df_4save.to_csv('similar_mock_gene_list.txt', sep='\\t',index = False)\n",
    "df_4save\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I suspect that isn't going to work in a reasonable amount of time as I couldn't even get to 21% similarity after running it for several minutes.\n",
    "Decided just to download the mock list and alter the list by hand in Sublime text. Save and upload for use in the demo.\n",
    "\n",
    "I found the code below useful when I was trying to make sure I didn't introduce any duplicates when I was editing by hand."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "l = similar_ranked_list\n",
    "z = set([x for x in l if l.count(x) > 1])\n",
    "print(z)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
